govcd/access_control.go

```diff
@@ -0,0 +1,230 @@
+func (vdc *Vdc) GetVappAccessControl(vappIdentifier string) (*types.ControlAccessParams, error) {
+	vapp, err := vdc.GetVAppByNameOrId(vappIdentifier, true)
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving vapp %s: %s", vappIdentifier, err)
+	}
+	return vapp.GetAccessControl()
+}
+
+// GetCatalogAccessControl is a convenience method to retrieve access control for a catalog
+// from an organization.
+// The input variable catalogIdentifier can be either the catalog name or its ID
+func (org *AdminOrg) GetCatalogAccessControl(catalogIdentifier string) (*types.ControlAccessParams, error) {
+	catalog, err := org.GetAdminCatalogByNameOrId(catalogIdentifier, true)
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving catalog %s: %s", catalogIdentifier, err)
+	}
+	return catalog.GetAccessControl()
+}
+
+// GetCatalogAccessControl is a convenience method to retrieve access control for a catalog
+// from an organization.
+// The input variable catalogIdentifier can be either the catalog name or its ID
+func (org *Org) GetCatalogAccessControl(catalogIdentifier string) (*types.ControlAccessParams, error) {
+	catalog, err := org.GetCatalogByNameOrId(catalogIdentifier, true)
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving catalog %s: %s", catalogIdentifier, err)
+	}
+	return catalog.GetAccessControl()
+}
+
+// GetAccessControl retrieves the access control information for this catalog
+func (catalog Catalog) GetAccessControl() (*types.ControlAccessParams, error) {
+
+	if catalog.Catalog.HREF == "" {
+		return nil, fmt.Errorf("catalog HREF is empty")
+	}
+	href := strings.Replace(catalog.Catalog.HREF, "/admin/", "/", 1)
+
+	return catalog.client.GetAccessControl(href, "catalog", catalog.Catalog.Name)
+}
+
+// SetAccessControl changes the access control information for this catalog
+func (catalog Catalog) SetAccessControl(accessControl *types.ControlAccessParams) error {
+
+	if catalog.Catalog.HREF == "" {
+		return fmt.Errorf("catalog HREF is empty")
+	}
+
+	href := strings.Replace(catalog.Catalog.HREF, "/admin/", "/", 1)
+
+	// When we set IsSharedToEveryone in the UI, what happens behind the scenes is that the request is changed
+	// to a read/only share to all visible organizations
+
+	if accessControl.IsSharedToEveryone {
+		return fmt.Errorf("share to everyone is not allowed for catalogs. You should set the sharing for all orgs that you need")
+	}
+
+	return catalog.client.SetAccessControl(accessControl, href, "catalog", catalog.Catalog.Name)
+}
+
+// RemoveAccessControl is a shortcut to SetAccessControl with all access disabled
+func (catalog Catalog) RemoveAccessControl() error {
+	return catalog.SetAccessControl(&types.ControlAccessParams{IsSharedToEveryone: false})
+}
+
+// IsShared shows whether a catalog is shared or not, regardless of the number of subjects sharing it
+func (catalog Catalog) IsShared() bool {
+	settings, err := catalog.GetAccessControl()
+	if err != nil {
```

Comment:
same questions as in previous functions.

Reply:
ditto